rainier-core/src/main/scala/com/stripe/rainier/compute/Target.scala

```diff
@@ -18,6 +18,45 @@ class Target(val real: Real, val placeholders: Map[Variable, Array[Double]]) {
 
   private def nRows(data: Map[Variable, Array[Double]]): Int =
     data.head._2.size
+
+  def batched(nBatches: Int): (Target, Real) =
+    if (placeholders.isEmpty)
+      (this, Real.zero)
+    else if (nBatches == 1)
+      (Target.empty, inlined)
+    else {
+      val n = nRows(placeholders)
+      val variables = placeholders.keys.toList
+      val nColumns = n / nBatches
```

Comment:
Okay, I think I get what we're doing, but I had to think more even after we talked it out. Let me know if this sounds right. By "an `n x m` matrix" I mean "a matrix with `n` rows and `m` columns."
So, placeholders is an `n x v` matrix with `n` rows, the lengths of the arrays and `v` columns, the number of variables. If `nBatches = b` and `m = n / b`, ignoring remainders for the moment, we are going to rearrange this into a `b x vm` matrix where we'll think of each row of this new matrix as a batch. We'll do this as follows: take our original `n x v` matrix and slice it horizontally into `m` slices each of which has `b` rows. This is already potentially confusing because the number of slices here is _not_ `nBatches = b`, it is `m`. Our slices are still stacked vertically, one on top of another. Now, we take these slices and place them side by side, left to right. By placing our `m` slices, each of dimension `b x v` next to each other, we get the `b x vm` matrix we want.

Let's do an example: suppose that we have a `6 x 2` matrix with entries `a_{ij}` (i = row, j = column) and we want `b = 2` batches. Then `m = 3`. Rearranging the slices, we get the matrix

```
a11  a12  a31  a32  a51  a52
a21  a22  a41  a42  a61  a62 
```

The first slice contains the first two rows and all of the columns, the next slice the next two rows and all of the columns, and so on. Here, `copies` makes copies of all of the variables and does so `m` times, one for each slice. To make the new placeholders, we iterate over the slices and take the coefficients from the original matrix which are in the slice we're in.

My only code comment so far is: I see that `n / nBatches` is the number of slice we are arranging side-by-side, but calling it `nColumsn` is pretty confusing. I think `nSlices` would be better.

Apologies for so much talking with so few helpful suggestions. I'll have more after lunch.

Reply:
Yes, that's correct. I agree that the naming could be better, I'll work on that before we merge. (FWIW, the idea behind "nColumns" was something more like "ihow many columns do we end up with *for each v*")